File: Python/Hard/3459_Length_of_Longest_V_Shaped_Diagonal_Segment/solution.py
class Solution:
    def lenOfVDiagonal(self, grid: list[list[int]]) -> int:
        directions = {
            "leftTop": (-1, -1),
            "leftBot": (1, -1),
            "rightTop": (-1, 1),
            "rightBot": (1, 1)
        }

        clockwise = {
            "leftTop": "rightTop",
            "rightTop": "rightBot",
            "rightBot": "leftBot",
            "leftBot": "leftTop"
        }

        maxCount = 0
        memo = {}
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == 1:
                    for d in directions:
                        dr, dc = directions[d]
                        maxCount = max(maxCount, 1 + Solution.explore(r + dr, c + dc, grid, 1, d, False, memo, directions, clockwise))
        return maxCount

    def explore(r, c, grid, prev, currentDirection, hasMadeTurn, memo, directions, clockwise):
        if not Solution.inbound(r, c, grid):
            return 0
        key = (r, c, prev, currentDirection, hasMadeTurn)
        if key in memo:
            return memo[key]
        
        if grid[r][c] == 1:
            return 0
        if prev == 1 and grid[r][c] != 2:
            return 0
        if prev == 2 and grid[r][c] != 0:
            return 0
        if prev == 0 and grid[r][c] != 2:
            return 0
        
        maxPath = 0
        dr, dc = directions[currentDirection]

        # go straight
        maxPath = max(maxPath, Solution.explore(r + dr, c + dc, grid, grid[r][c], currentDirection, hasMadeTurn, memo, directions, clockwise))

        # turn
        if not hasMadeTurn:
            newDirection = clockwise[currentDirection]
            dr, dc = directions[newDirection]
            maxPath = max(maxPath, Solution.explore(r + dr, c + dc, grid, grid[r][c], newDirection, True, memo, directions, clockwise))
        
        memo[key] = maxPath + 1
        return memo[key]


    def inbound(r, c, grid):
        rInbound = r >= 0 and r < len(grid)
        cInbound = c >= 0 and c < len(grid[0])
        return rInbound and cInbound

File: Golang/Hard/3459_Length_of_Longest_V_Shaped_Diagonal_Segment/solution.go
package lengthoflongestvshapeddiagonalsegment

type state struct {
	r                int
	c                int
	prev             int
	currentDirection string
	hasMadeTurn      bool
}

func lenOfVDiagonal(grid [][]int) int {
	directions := map[string][]int{
		"leftTop":  {-1, -1},
		"leftBot":  {1, -1},
		"rightTop": {-1, 1},
		"rightBot": {1, 1},
	}

	clockwise := map[string]string{
		"leftTop":  "rightTop",
		"rightTop": "rightBot",
		"rightBot": "leftBot",
		"leftBot":  "leftTop",
	}

	maxCount := 0
	memo := make(map[state]int)

	for r := range grid {
		for c := range grid[0] {
			if grid[r][c] == 1 {
				for d, _ := range directions {
					dr := directions[d][0]
					dc := directions[d][1]
					maxCount = max(maxCount, 1+explore(r+dr, c+dc, grid, 1, d, false, directions, clockwise, memo))
				}
			}
		}
	}
	return maxCount
}

func explore(r int, c int, grid [][]int, prev int, currentDirection string, hasMadeTurn bool, directions map[string][]int, clockwise map[string]string, memo map[state]int) int {
	if !inbound(r, c, grid) {
		return 0
	}

	key := state{r, c, prev, currentDirection, hasMadeTurn}
	if v, exist := memo[key]; exist {
		return v
	}

	if grid[r][c] == 1 {
		return 0
	}

	if prev == 1 && grid[r][c] != 2 {
		return 0
	}

	if prev == 2 && grid[r][c] != 0 {
		return 0
	}

	if prev == 0 && grid[r][c] != 2 {
		return 0
	}

	maxPath := 0
	dr := directions[currentDirection][0]
	dc := directions[currentDirection][1]

	maxPath = max(maxPath, explore(r+dr, c+dc, grid, grid[r][c], currentDirection, hasMadeTurn, directions, clockwise, memo))

	if !hasMadeTurn {
		newDirection := clockwise[currentDirection]
		dr := directions[newDirection][0]
		dc := directions[newDirection][1]
		maxPath = max(maxPath, explore(r+dr, c+dc, grid, grid[r][c], newDirection, true, directions, clockwise, memo))
	}

	memo[key] = maxPath + 1
	return memo[key]
}

func inbound(r int, c int, grid [][]int) bool {
	rInbound := r >= 0 && r < len(grid)
	cInbound := c >= 0 && c < len(grid[0])
	return rInbound && cInbound
}
